pulp_docker/app/tasks/stages/dedupe_save.py

```diff
@@ -0,0 +1,163 @@
+from django.db.models import Q
+from pulpcore.plugin.stages import Stage
+from pulpcore.plugin.models import Artifact, ContentArtifact, RemoteArtifact
+# from pulpcore.plugin.models import ProgressBar
+
+import logging
+log = logging.getLogger("STUPIDSAVE")
+
+
+class SerialArtifactSave(Stage):
+    """
+    Save Artifacts one at a time, combining duplicates.
+    """
+
+    async def __call__(self, in_q, out_q):
+        """
+        The coroutine for this stage.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+        Returns:
+            The coroutine for this stage.
+
+        """
+        while True:
+            dc = await in_q.get()
+            if dc is None:
+                break
+            self.query_and_save_artifacts(dc)
+            await out_q.put(dc)
+        await out_q.put(None)
+
+    def query_and_save_artifacts(self, dc):
+        """
+        Combine duplicated artifacts, save unique artifacts.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+        """
+        for da in dc.d_artifacts:
+            artifact_q = Q()
+            for digest_name in da.artifact.DIGEST_FIELDS:
+                digest_value = getattr(da.artifact, digest_name)
+                if digest_value:
+                    key = {digest_name: digest_value}
+                    artifact_q &= Q(**key)
+            try:
+                existing_artifact = Artifact.objects.get(artifact_q)
+                # TODO count deduped artifacts
+            except Artifact.DoesNotExist as e:
+                da.artifact.save()
+            else:
+                da.artifact = existing_artifact
+
+
+class SerialContentSave(Stage):
+    """
+    Combine duplicated Content, save unique Content.
+    """
+
+    async def __call__(self, in_q, out_q):
+        """
+        The coroutine for this stage.
+
+        Args:
+            in_q (:class:`asyncio.Queue`): The queue to receive
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` objects from.
+            out_q (:class:`asyncio.Queue`): The queue to put
+                :class:`~pulpcore.plugin.stages.DeclarativeContent` into.
+        Returns:
+            The coroutine for this stage.
+
+        """
+        while True:
+            dc = await in_q.get()
+            # finished
+            if dc is None:
+                break
+            # Artifacts have not been downloaded
+            elif not self.settled(dc):
```

Comment:
This can just be `if`.  The `elif` here is a little confusing.

Pretty sure the preceeding `ArtifactDownload` will only output the DC after all the artifacts are downloaded so don't see the value in the `settled` test here.  Also, when downloading is deferred (Lazy), the `Artifact` will never be saved.

Reply:
You are right, but after those artifacts are saved and processed, there will be new nested units created by the ProcessContent Stage. These won't have been downloaded yet-- so this is one of the complexities of repeating a single stage to handle units in various states. For lazy, we could only call self.settled on ManifestLists and Manifests.

Reply:
Just to be clear, `settled` is still in there, but will be removed when we refactor to use the updated pulpcore stages that now use the custom manager.